### DISCUSSION TITLE
How to disable the welcome page in Keycloak 17

### DISCUSSION THREAD

**Discussion by hdava** [keycloak/keycloak#10467] [Archive]
The new distribution of Keycloak that is powered by Quarkus does not have a standalone.xml file in order to achieve this purpose.

How can I disable the path and (if needed) how can I customize this page?

**Comment by stianst** [comment]
Could you mention how you disable it in standalone.xml? Is it with a Undertow filter perhaps?

**Comment by stianst** [comment]
On Quarkus we'd probably want this to be exposed as a feature, something like `--features-disabled=welcome`.

**Reply by MohammedNoureldin** [reply]
Hi, do we need actually to "hide" or add an additional protection layer using reverse proxy to the admin login page? Or is it secure enough just to have it "public", at least, to those who have access to the internal network.

**Comment by apiraino** [comment]
I was also surprised that there is no straightforward way to disable that page, certainly they shouldn't be public in a production deployment.

On Keycloak 17.0 I've tried theming that page with a `./my-custom-theme/welcome/info.ftl` file and override [this file](https://github.com/keycloak/keycloak/tree/main/themes/src/main/resources/theme/keycloak/welcome), but I didn't manage to make it work.

As a last resort, I've configured on the reverse proxy an authentication layer for the welcome page and the admin login.

I'd love to learn a less hacky way to fix this, if there's one.

**Reply by MohammedNoureldin** [reply]
Hi, are there any news regarding this?

**Comment by nicolas-goudry** [comment]
Personnally I customized the theme of the welcome page to automatically redirect to the admin authentication page (like if you clicked on the link to launch the admin console)… Not ideal, but at least the welcome page is not browsable.

**theme/welcome/theme.properties**
```properties
import=common/keycloak
locales=en
```

**theme/welcome/index.ftl**
```ftl
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="refresh" content="0; url=${adminUrl}" />
    <meta name="robots" content="noindex, nofollow" />
    <script type="text/javascript">
      window.location.href = '${adminUrl}'
    </script>
  </head>
  <body>
    If you are not redirected automatically, follow this <a href="${adminUrl}">link</a>.
  </body>
</html>
```

Then, I defined the `KC_SPI_THEME_WELCOME_THEME` variable to the name of my custom theme.

I did it the same way for Keycloak 16, the only difference is the variable name that changed. It was previously named `KEYCLOAK_WELCOME_THEME`.

**Comment by us4r93t3mnus** [comment]
Hello, any updates for this issue? I can't deal with locales on welcome pages... Any help?

**Comment by ssilvert** [comment]
@us4r93t3mnus  In Keycloak 24, the Welcome page is only used to create the initial admin user.  After that, any time you try to go to the Welcome page it will redirect to the admin console and force a login.

Does that help?

**Reply by us4r93t3mnus** [reply]
That's true. 

I was facing the welcome page as an index webpage and I was wrong. Thanks for the info.

**Reply by JeanBilski** [reply]
Hi! That's a good start. An ideal solution would be to be able to define a redirect url for the users to be redirected to I think, like nicolas-goudry did. As an exemple we want to redirect our users on our main app when they are hitting the welcome page of keycloak, to trigger the auth flow if they are disconnected. If Keycloak would redirect to the admin login they would face a bad gateway as we are blocking the admin endpoint to internet

**Reply by Sieboldianus** [reply]
I am also looking into this. Did anyone succeed in overriding the welcome page with a redirect to the login-flow for another realm in 24.0.2, not the master/admin console? There's a `redirectToAdmin=true` in the default `theme.properties`. Maybe setting this to False and using @nicolas-goudry's approach could work? I will give it a try..

**Reply by Sieboldianus** [reply]
Yes, looks like it works. Added the steps [here](https://stackoverflow.com/questions/78251173/keycloak-how-to-override-welcome-screen-redirect-behavior-to-custom-realm-ins/78253866#78253866)

**Reply by mkopeyka** [reply]
I used redirect in `welcome` theme in v20.0.2 to redirect to specific realm but v25 ignores redirect in `index.ftl` and redirects me to `master` realm login page.
```
<!DOCTYPE html>
<html lang="en">
<head>
    <title>Welcome to ${productName}</title>
    <meta charset="utf-8">
    <meta http-equiv="refresh" content="0; url=/auth" />
    <meta name="robots" content="noindex, nofollow">
    <script type="text/javascript">
        window.location.href = "/auth/admin/my-realm/console"
    </script>
</head>
  <body>
    <p>Please follow <a href="/auth/admin/my-realm/console">this link</a> to open my-realm page.</p>
  </body>
</html>
```

**Reply by Sieboldianus** [reply]
You have to specify the realm name. Also maybe add the baseUrl. This is what I use to redirect users to the Account Console:
```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="refresh" content="0; url=${baseUrl}realms/${properties.realmForwardTo}/account/#/" />
    <meta name="robots" content="noindex, nofollow" />
    <script type="text/javascript">
      window.location.href = '${baseUrl}realms/${properties.realmForwardTo}/account/#/'
    </script>
  </head>
  <body>
    If you are not redirected automatically, follow this <a href="${baseUrl}realms/${properties.realmForwardTo}/account/#/">link</a>.
  </body>
</html>

```

**Reply by mkopeyka** [reply]
I tried absolute url to admin console for my-realm, the same situation, keycloak 25.0.1 redirects to `master` realm
```
<!DOCTYPE html>
<html lang="en">
<head>
    <title>Welcome to ${productName}</title>
    <meta charset="utf-8">
    <meta http-equiv="refresh" content="0; url=http://localhost:9090/auth/admin/my-realm/console/" />
    <meta name="robots" content="noindex, nofollow">
    <script type="text/javascript">
        window.location.href = "http://localhost:9090/auth/admin/my-realm/console/"
    </script>
</head>
  <body>
    <p>Please follow <a href="http://localhost:9090/auth/admin/my-realm/console/">this link</a> to open my-realm page.</p>
  </body>
</html>
```
